pkg/blockbuilder/blockbuilder.go

```diff
@@ -371,13 +361,13 @@ func (b *BlockBuilder) consumePartitions(ctx context.Context, part int32, mark t
 				// 2. process was restarted, the client fetched last commited offset for the group+partition and started consuming from this (old) offset
 				recordProcessedBefore := false // TODO(codesome): get this from checkpoint
 				allSamplesProcessed, err := builder.process(ctx, rec, lastEnd, currEnd, recordProcessedBefore)
-				if !allSamplesProcessed && checkpointOffset < 0 {
-					checkpointOffset = rec.Offset
-				}
+				_ = allSamplesProcessed
 				if err != nil {
 					level.Error(b.logger).Log("msg", "failed to process record", "part", part, "key", string(rec.Key), "err", err)
 					// TODO(codesome): do we just ignore this? What if it was Mimir's issue and this leading to data loss?
 					// TODO(codesome): add metric
+				} else {
+					checkpointRec = rec
```

Comment:
According to our design, this might not be right. The checkpoint tells all samples before this has been processed. Some samples in some records may not be processed. So we should make use of `allSamplesProcessed` to determine that.

Reply:
I think, the `allSamplesProcessed` needs revisiting in general (refer to the [TODO about that above][1]). The problem right now is that if `builder.process` skips samples, these samples will be lost. This is because on the next cycle, Kafka client will continue consuming from its internal offset, not from the checkpoint. Recall that consumption in Kafka only moves forward, checkpointing is needed only for the case, when process was restarted (or rebalance happens and another client takes over the partition).

[1]: https://github.com/grafana/mimir/blob/0c290c7061034564b5ec9f867f22078332b47f18/pkg/blockbuilder/blockbuilder.go#L358-L361

Reply:
Yup that is a problem we need to fix and I plan to take a look at it either tomorrow or next week. But the meaning of commit would remain the same I believe where everything before that is in a block. It is how we consume needs updating. We can consider starting from the last commit everytime we compact and close a builder (not the most efficient, but a start).

I would either add a TODO here to fix the commit point or fix the commit point and take care of proper consumption as a next task (I would prefer the latter).

Reply:
Question: If we commit at this record, will we start at this record when we read from a commit next time or do we start from the record after this?

Reply:
I don't think I got the question, but to point this out, here "commit a record" is only a shortcut implemented by Kafka client. It does the following:

1. take the topic of the record
2. take the partition of the record
3. take the offset of the record
4. commit the tuple `(topic, partition, offset)`

Refer to the [implementation](https://github.com/twmb/franz-go/blob/v1.17.0/pkg/kgo/consumer_group.go#L2349).

Kafka client will start its next `PollFetches` from the last offset it consumed, unless the consumption process restarted (e.g. rebalanced or crashed).